### PR TITLE
multi: allow custom rejection messages in ChannelAcceptor

### LINKED ISSUE #3552: Channel acceptor should allow for custom rejection message
### Background

When dynamically rejecting an inbound channel, the connecting peer should receive an indication of why their request was rejected.

Examples of existing messages:

1. Minimum channel size not reached
2. Multiple channels not supported

### Steps to reproduce

1. Open a channel that will get dynamically rejected
2. Get generic rejection message

### Expected behaviour

Rejection message should carry a hint at the problem for debugging purposes

### Actual behaviour

Message cannot be customized so the receiver doesn't know why the channel was rejected

### PR DESCRIPTION
**Pull request by Jeiwan**
This PR fixes #3552 . It:
1. Adds `RejectionReason` field to `ChannelAcceptResponse` message. `Accept` field is still used to check if request was accepted/rejected, thus `RejectionReason` is kind of optional and can be empty.
2. Changes `Accept` method of `ChannelAcceptor` interface to return an error instead of a bool.
3. Changes `Accept` method of `ChainedAcceptor` to also return an error. If there are multiple acceptors that would reject a request, first error will be returned.
4. Adds `chanacceptor.AcceptanceError` that wraps rejection error message.
5. Modifies `fundingmanager` to wrap errors returned by `OpenChannelPredicate` in `AcceptanceError` and send them to peers.
6. If `RejectionReason` wasn't set, a general error is sent to the peer.

#### Pull Request Checklist

- [x] If this is your first time contributing, we recommend you read the [Code
  Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md)
- [x] All changes are Go version 1.12 compliant
- [x] The code being submitted is commented according to [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation)
- [x] For new code: Code is accompanied by tests which exercise both
  the positive and negative (error paths) conditions (if applicable)
- [x] Any new logging statements use an appropriate subsystem and
  logging level
- [x] Code has been formatted with `go fmt`
- [x] Protobuf files (`lnrpc/**/*.proto`) have been formatted with
  `make rpc-format` and compiled with `make rpc`
- [x] For code and documentation: lines are wrapped at 80 characters
  (the tab character should be counted as 8 characters, not 4, as some IDEs do
  per default)
- [x] Running `make check` does not fail any tests
- [x] Running `go vet` does not report any issues
- [x] Running `make lint` does not report any **new** issues that did not
  already exist
- [x] All commits build properly and pass tests. Only in exceptional
  cases it can be justifiable to violate this condition. In that case, the
  reason should be stated in the commit message.
- [x] Commits have a logical structure according to [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure)

### CODE REVIEW

**Comment by Jeiwan** [comment]
Ran it locally and it worked good. I'll try to add some more tests and it'll be done.

**Comment by Jeiwan** [comment]
I don't see how to improve this any further.

**Comment by Jeiwan** [comment]
I'm still working on this, there are multiple things to improve.

**Comment by Jeiwan** [comment]
I think it's not a good idea to allow node operators to define the whole message because these messages might be used to confuse remote peers (when shown in UI, for example). So I made them use such template: `open channel request rejected: <custom message>`. It makes it clear what exactly happened.

I also limited the size of custom messages when they're sent to remote peers: only first 128 bytes (not including the prefix message) will be sent (this is just a random number that looks long enough, let me know if you think there's a better limit). Rejecting node will have the whole error message printed in its logs.

File: lnrpc/rpc.proto [hunk 1]
```diff
@@ -1105,6 +1105,9 @@ message ChannelAcceptResponse {
 
     /// The pending channel id to which this response applies.
     bytes pending_chan_id = 2;
+
+    /// The message explaining why the channel was rejected.
+    string rejectionReason = 3;
```
[the remark this answers is not in the document]
Comment (Jeiwan) [reply lnrpc/rpc.proto]: Fixed

File: chanacceptor/acceptor_test.go [hunk 2]
```diff
@@ -110,15 +110,19 @@ func TestRPCMultipleAcceptClients(t *testing.T) {
 			pendingID := req.OpenChanMsg.PendingChannelID
 			if !bytes.Equal(pendingID[:], resp.PendingChanId) {
 				errChan <- struct{}{}
-				return false
+				return errors.New("PendingCanId doesn't match the ID in ChannelAcceptRequest")
```
[the remark this answers is not in the document]
Comment (Jeiwan) [reply chanacceptor/acceptor_test.go]: Fixed

File: lnrpc/rpc.proto [hunk 3]
```diff
@@ -1105,6 +1105,9 @@ message ChannelAcceptResponse {
 
     /// The pending channel id to which this response applies.
     bytes pending_chan_id = 2;
+
+    /// The message explaining why the channel was rejected.
```
[the remark this answers is not in the document]
Comment (Jeiwan) [reply lnrpc/rpc.proto]: I agree with adding a note. Fixed.

Yep, having separate internal and external errors looks like overkill. If needed, we can allow to opt out of sending the custom message to the remote peer (send the default one instead) and just print it to logs.

File: chanacceptor/chainedacceptor_test.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	allowingAcceptor := customAcceptor{}
+	rejectingAcceptor := unwantedNodeRejector{nodeB}
+	rejectingAcceptor2 := unwantedChannelRejector{firstReq.OpenChanMsg.PendingChannelID}
+
+	tests := []struct {
+		acceptors []ChannelAcceptor
+		request   *ChannelAcceptRequest
+		result    error
+	}{
+		{acceptors: []ChannelAcceptor{allowingAcceptor}, request: firstReq, result: nil},
```
[the remark this answers is not in the document]
Comment (Jeiwan) [reply chanacceptor/chainedacceptor_test.go]: Fixed

**Comment by Jeiwan** [comment]
Rebased.

**Comment by Jeiwan** [comment]
Rebased.

File: chanacceptor/chainedacceptor_test.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+	}
+
+	for _, test := range tests {
+		chainedAcceptor := NewChainedAcceptor()
+
+		for _, acceptor := range test.acceptors {
+			chainedAcceptor.AddAcceptor(acceptor)
+		}
+
+		got := chainedAcceptor.Accept(test.request)
```
Comment (Crypt-iQ) [inline chanacceptor/chainedacceptor_test.go:204]: confusing variable name, maybe `result` instead?

File: rpcserver.go [hunk 6]
```diff
@@ -6059,8 +6061,9 @@ func (r *rpcServer) ChannelAcceptor(stream lnrpc.Lightning_ChannelAcceptorServer
 			copy(pendingID[:], resp.PendingChanId)
 
 			openChanResp := lnrpc.ChannelAcceptResponse{
-				Accept:        resp.Accept,
-				PendingChanId: pendingID[:],
+				Accept:          resp.Accept,
+				PendingChanId:   pendingID[:],
+				RejectionReason: resp.RejectionReason,
```
Comment (Crypt-iQ) [inline rpcserver.go:6068]: `RejectionReason` not yet added, code will fail to compile

File: rpcserver.go [hunk 7]
```diff
@@ -6022,13 +6023,14 @@ func (r *rpcServer) ChannelAcceptor(stream lnrpc.Lightning_ChannelAcceptorServer
...
 		case <-timeout:
-			rpcsLog.Errorf("RPCAcceptor returned false - reached timeout of %d",
+			err := fmt.Errorf("RPCAcceptor returned false - reached timeout of %d",
 				r.cfg.AcceptorTimeout)
-			return false
+			rpcsLog.Error(err)
+			return err
 		case <-quit:
-			return false
+			return errors.New("RPCAcceptor terminated")
```
Comment (Crypt-iQ) [inline rpcserver.go:6033]: ideally use an exported error from `chanacceptor` package

File: chanacceptor/errors.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+// A compile time check to ensure AcceptanceError implements the error
+// interface.
+var _ error = (*AcceptanceError)(nil)
+
+// ErrRejected returns a general open channel error. It is sent to a remote node
+// when a custom ChannelAcceptor rejected an open channel request and didn't
+// specify any rejection message.
+func ErrRejected() *AcceptanceError {
```
Comment (Crypt-iQ) [inline chanacceptor/errors.go]: should be `AcceptanceError` not `*AcceptanceError`

File: chanacceptor/errors.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+	return &AcceptanceError{
+		errors.New("open channel request rejected"),
+	}
+}
+
+// ErrRejectedWithMsg returns a custom open channel error. It is used by the
+// funding manager to wrap and send rejection error messages to remote nodes.
+func ErrRejectedWithMsg(rejectionMsg string) *AcceptanceError {
```
[the remark this answers is not in the document]
Comment (Crypt-iQ) [reply chanacceptor/errors.go]: Agree, even then I'm not sure how necessary the "open channel request received" prefix is...

File: fundingmanager.go [hunk 10]
```diff
@@ -1272,9 +1275,15 @@ func (f *fundingManager) handleFundingOpen(fmsg *fundingOpenMsg) {
 	}
 
 	if err := f.cfg.OpenChannelPredicate.Accept(chanReq); err != nil {
+		err = chanacceptor.ErrRejectedWithMsg(err.Error())
+
+		if err.Error() == "" {
+			err = chanacceptor.ErrRejected()
```
Comment (Crypt-iQ) [inline fundingmanager.go:1281]: See comment on 1903da6 about combining `ErrRejected` with `ErrRejectedWithMsg`

File: chanacceptor/chainedacceptor_test.go [hunk 11]
```diff
@@ -72,52 +72,128 @@ func TestChainedAcceptorRejectingAcceptor(t *testing.T) {
...
 			},
-			request: firstReq,
-			result:  errUnwantedChannel},
+			request: reqNodeA,
+			result:  errUnwantedChannel,
+		},
 	}
 
 	for _, test := range tests {
+		t.Logf("Running test case: %v", test.name)
```
[the remark this answers is not in the document]
Comment (Crypt-iQ) [reply chanacceptor/chainedacceptor_test.go:196]: 👍 also this commit should be a fixup of ab9a505

File: chanacceptor/errors.go [hunk 12]
```diff
@@ -30,6 +31,6 @@ func ErrRejected() AcceptanceError {
 // funding manager to wrap and send rejection error messages to remote nodes.
 func ErrRejectedWithMsg(rejectionMsg string) AcceptanceError {
 	return AcceptanceError{
-		errors.New(rejectionMsg),
+		fmt.Errorf("open channel request rejected: %s", rejectionMsg),
```
Comment (Crypt-iQ) [inline chanacceptor/errors.go:52]: should be a fixup of 1903da6

File: fundingmanager_test.go [hunk 13]
```diff
@@ -2384,6 +2392,126 @@ func TestFundingManagerPrivateRestart(t *testing.T) {
...
+			t.Fatalf("expected OpenChannel to be sent from "+
+				"alice, instead got %T", aliceMsg)
+		}
+
+		// Let Bob handle the init message.
+		bob.fundingMgr.processFundingOpen(openChannelReq, alice)
+
+		// Assert Bob responded.
+		// First two requests should be rejected.
+		if i < 2 {
```
Comment (Crypt-iQ) [inline fundingmanager_test.go]: this can just be a bool that is unrelated to the position in the array. A field in the test called `expectBobRespond` for example

File: fundingmanager_test.go [hunk 14]
```diff
@@ -2384,6 +2392,126 @@ func TestFundingManagerPrivateRestart(t *testing.T) {
...
+	if len(rejected) != 2 {
+		t.Fatalf("expected to get 2 rejects, got %d", len(rejected))
+	}
+
+	if len(accepted) != 1 {
+		t.Fatalf("expected to get 1 request accepted, got %d", len(accepted))
+	}
+
+	// First reject should be the default open channel rejection error.
+	err := rejected[0]
```
[the remark this answers is not in the document]
Comment (Crypt-iQ) [reply fundingmanager_test.go:2517]: 👍

File: chanacceptor/acceptor_test.go [hunk 15]
```diff
@@ -141,8 +145,9 @@ func TestRPCMultipleAcceptClients(t *testing.T) {
 		select {
 		case newRequest := <-requests:
 			newResponse := lnrpc.ChannelAcceptResponse{
-				Accept:        true,
-				PendingChanId: newRequest.chanReq.OpenChanMsg.PendingChannelID[:],
+				Accept:          true,
+				PendingChanId:   newRequest.chanReq.OpenChanMsg.PendingChannelID[:],
+				RejectionReason: "",
```
[the remark this answers is not in the document]
Comment (Crypt-iQ) [reply chanacceptor/acceptor_test.go:153]: This commit should be squashed with the commit it fixes

File: chanacceptor/errors.go [hunk 16]
```diff
@@ -5,15 +5,32 @@ import (
 	"fmt"
 )
 
+const (
+	// Length of the prefix message in `ErrRejectedWithMsg` + the max. length of
+	// custom message allowed to be sent to the remote peer.
+	trimmedAcceptanceErrorLength = 31 + 128
```
Comment (Crypt-iQ) [inline chanacceptor/errors.go:11]: This commit should probably be in the commit that introduces the AcceptanceError. Otherwise the diff gets larger b/t commits because of test changes.

File: fundingmanager_test.go [hunk 17]
```diff
@@ -2384,6 +2392,156 @@ func TestFundingManagerPrivateRestart(t *testing.T) {
...
+		t.Fatalf("expected ErrRejectedWithErr error, got \"%v\"",
+			err.Error())
+	}
+
+	// Third reject should be a trimmed open channel rejection error.
+	err = rejected[2]
+	rejectionMsg := strings.Repeat("a", 128)
+	if !strings.HasSuffix(
+		err.Error(),
+		"open channel request rejected: "+rejectionMsg+"...") {
```
[the remark this answers is not in the document]
Comment (Crypt-iQ) [reply fundingmanager_test.go:2538]: Not convinced we need the prefix. Also should squash this commit with a relevant one since this commit is changing a different commit you made.

**Comment by carlaKC** [comment]
Needs rebase!

**Comment by Roasbeef** [comment]
The functionality of this PR has been folded into: #4746 4683
